icu4c/source/i18n/currpinf.cpp

```diff
@@ -296,18 +285,18 @@ CurrencyPluralInfo::setupCurrencyPluralPattern(const Locale& loc, UErrorCode& st
         return;
     }
     UErrorCode ec = U_ZERO_ERROR;
-    UResourceBundle *rb = ures_open(nullptr, loc.getName(), &ec);
-    UResourceBundle *numElements = ures_getByKeyWithFallback(rb, gNumberElementsTag, nullptr, &ec);
-    rb = ures_getByKeyWithFallback(numElements, ns->getName(), rb, &ec);
-    rb = ures_getByKeyWithFallback(rb, gPatternsTag, rb, &ec);
+    LocalUResourceBundlePointer rb(ures_open(nullptr, loc.getName(), &ec));
+    LocalUResourceBundlePointer numElements(ures_getByKeyWithFallback(rb.getAlias(), gNumberElementsTag, nullptr, &ec));
+    LocalUResourceBundlePointer numSystemScript(ures_getByKeyWithFallback(numElements.getAlias(), ns->getName(), nullptr, &ec));
```

Comment:
It's fine, but I think it should still be safe to use the third argument of `ures_getByKeyWithFallback` rather than making a new LocalUResourceBundlePointer for every level in the resource bundle tree.

Reply:
I thought about that too, but then I was wondering who would call close on the ResourceBundle? (and in the correct order).
IIUC, calling `orphan()` doesn't call the close function, and if you call `getAlias()`, then the original bundle might be overwritten before the `adoptInstead()` can call close.

Ex:
 rb.adoptInstead(ures_getByKeyWithFallback(numElements.getAlias(), ns->getName(), rb.orphan(), &ec));

 rb.adoptInstead(ures_getByKeyWithFallback(numElements.getAlias(), ns->getName(), rb.getAlias(), &ec));


---
In reply to: [207658933](https://github.com/unicode-org/icu/pull/17#discussion_r207658933) [](ancestors = 207658933)

Reply:
What I meant was, I think it is safe to do,

```
LocalUResourceBundlePointer rb(ures_open(...));
ures_getByKeyWithFallback(rb.getAlias(), "...", rb.getAlias(), &ec);
ures_getByKeyWithFallback(rb.getAlias(), "...", rb.getAlias(), &ec);
ures_getByKeyWithFallback(rb.getAlias(), "...", rb.getAlias(), &ec);
```

Example of a place in ICU code that uses this pattern already:

https://github.com/unicode-org/icu/blob/778d0a6d1d46faa724ead19613bda84621794b72/icu4c/source/i18n/reldtfmt.cpp#L450

I'm double-checking with Markus that this is safe.

Reply:
After studying [uresbund.cpp](https://github.com/unicode-org/icu/blob/778d0a6d1d46faa724ead19613bda84621794b72/icu4c/source/common/uresbund.cpp#L1212), I can't find any code paths in `ures_getByKeyWithFallback` and related functions that *obviously* return a different object than the one you passed via the fillIn parameter.  Markus didn't have an answer.  The documentation also does not provide any clues.

There is one thing that we can assume, and it's that when you pass a fillIn parameter to `ures_getByKeyWithFallback`, that you are passing ownership to that function, because most call sites, including the one you are refactoring, overwrite the original `rb` pointer with the return value.  In most cases, `ures_getByKeyWithFallback` hands ownership of fillIn right back again by returning the same pointer; the question which I can't give a 100% certain answer to is whether fillIn is *always* the return value when it is non-null.

What I can see by looking at the code is that you should be able to expect better performance by using the fillIn, because without the fillIn, you add extra mallocs.

I think the pattern I posted above is probably fine, since I can find other places in ICU code that use it, but if you are hesitant, then I think the following pattern is also definitely safe:

```
LocalUResourceBundlePointer rb(ures_open(...));
rb.adoptInstead(ures_getByKeyWithFallback(rb.getAlias(), "...", rb.orphan(), &ec));
rb.adoptInstead(ures_getByKeyWithFallback(rb.getAlias(), "...", rb.orphan(), &ec));
rb.adoptInstead(ures_getByKeyWithFallback(rb.getAlias(), "...", rb.orphan(), &ec));
```

`rb.getAlias()` gets called first, then `rb.orphan()`, then `ures_getByKeyWithFallback` does its stuff, then you save the result back into `rb`.

Reply:
Thanks Shane.
My hesitation about using that exact pattern was more about the question of "who calls the  `ures_close`" once the code is done with the Resource Bundle, in order to ensure that there are no leaks. I was thinking that when you pass the fillIn it might not close.

However, from taking a close look at the code, it looks like it will automatically call the appropriate close methods if the `fillIn` argument is not `NULL`.
Inside the `init_resb_result` function:
```c++
    if(resB == NULL) {
        // .... omitted...
    } else {
        if(resB->fData != NULL) {
            entryClose(resB->fData);
        }
        if(resB->fVersion != NULL) {
            uprv_free(resB->fVersion);
        }
        // .... omitted...
    }
```

So, in short, yes it does seem to be safe to use this pattern.
Thanks Shane.